File: LeetCode/1622. Fancy Sequence/solution.py
class Fancy:

    def __init__(self):
        self.arr, self.add, self.mul = [], [0], [1]


    def append(self, val: int) -> None:
        self.arr.append(val)
        self.add.append(self.add[-1])
        self.mul.append(self.mul[-1])


    def addAll(self, inc: int) -> None:
        self.add[-1] += inc


    def multAll(self, m: int) -> None:
        self.mul[-1] *= m 
        self.add[-1] *= m


    def getIndex(self, idx: int) -> int:
        if idx >= len(self.arr):
            return -1
        # print(self.arr,self.add,self.mul)
        m, inc = self.mul[-1], self.add[-1]
        m //= self.mul[idx]
        inc -= m*self.add[idx]
        # print(m,inc)
        return (self.arr[idx]*m + inc)%(10**9+7)



# Your Fancy object will be instantiated and called as such:
# obj = Fancy()
# obj.append(val)
# obj.addAll(inc)
# obj.multAll(m)
# param_4 = obj.getIndex(idx)



class Fancy:

    def __init__(self):
        self.arr, self.add, self.mul = [], 0, 1
        self.MOD = 10**9+7


    def append(self, val: int) -> None:
        val -= self.add
        val %= self.MOD
        self.arr.append((val,self.mul))

    def addAll(self, inc: int) -> None:
        self.add += inc
        self.add %= self.MOD


    def multAll(self, m: int) -> None:
        self.mul *= m 
        self.add *= m
        # self.mul %= self.MOD
        self.add %= self.MOD


    def getIndex(self, idx: int) -> int:
        if idx >= len(self.arr):
            return -1
        # print(self.arr,self.add,self.mul)
        # print(m,inc)
        m, inc = self.mul, self.add
        a, b = self.arr[idx]
        return (a*m//b + inc)%self.MOD



# Your Fancy object will be instantiated and called as such:
# obj = Fancy()
# obj.append(val)
# obj.addAll(inc)
# obj.multAll(m)
# param_4 = obj.getIndex(idx)


class Fancy:

    def __init__(self):
        self.arr, self.add, self.mul = [], 0, 1
        self.MOD = 10**9+7

    def quick_mul(self, a, m):
        ret = 1
        while m:
            if m&1:
                ret *= a 
                ret %= self.MOD 
            a *= a 
            a %= self.MOD
            m = (m >> 1)
        return ret 

    def append(self, val: int) -> None:
        val -= self.add
        val = (val*self.quick_mul(self.mul, self.MOD-2))%self.MOD
        self.arr.append(val)

    def addAll(self, inc: int) -> None:
        self.add += inc
        self.add %= self.MOD


    def multAll(self, m: int) -> None:
        self.mul *= m 
        self.add *= m
        self.mul %= self.MOD
        self.add %= self.MOD


    def getIndex(self, idx: int) -> int:
        if idx >= len(self.arr):
            return -1
        # print(self.arr,self.add,self.mul)
        # print(m,inc)
        a, m, inc = self.arr[idx], self.mul, self.add
        return (a*m + inc)%self.MOD



# Your Fancy object will be instantiated and called as such:
# obj = Fancy()
# obj.append(val)
# obj.addAll(inc)
# obj.multAll(m)
# param_4 = obj.getIndex(idx)

File: LeetCode/1622. Fancy Sequence/solution.cpp
#define LL long long

class Fancy {
private:
    vector<LL> nums;
    LL add, mul;
    const int mod = 1000000007;
    
    LL power(LL x, int y) {
        LL tot = 1, p = x;
        for (; y; y >>= 1) {
            if (y & 1)
                tot = (tot * p) % mod;
            p = (p * p) % mod;
        }
        return tot;
    }

public:
    Fancy() {
        add = 0;
        mul = 1;
    }
    
    void append(int val) {
        val = ((val - add) % mod + mod) % mod;
        val = (val * power(mul, mod - 2)) % mod;
        nums.push_back(val);
    }
    
    void addAll(int inc) {
        add = (add + inc) % mod;
    }
    
    void multAll(int m) {
        add = add * m % mod;
        mul = mul * m % mod;
    }
    
    int getIndex(int idx) {
        if (idx >= nums.size())
            return -1;
        return (nums[idx] * mul + add) % mod;
    }
};

/**
 * Your Fancy object will be instantiated and called as such:
 * Fancy* obj = new Fancy();
 * obj->append(val);
 * obj->addAll(inc);
 * obj->multAll(m);
 * int param_4 = obj->getIndex(idx);
 */



using ll = long long;

const int MOD = 1e9+7;

class Fancy {
public:
    vector<ll> ks, bs, nums;
    ll k, b;
    Fancy() {
        k = 1, b = 0;
    }
    
    void append(int val) {
        nums.push_back(val);
        ks.push_back(k);
        bs.push_back(b);
    }
    
    void addAll(int inc) {
        b = (b + inc)%MOD;
    }
    
    void multAll(int m) {
        b = (b*m)%MOD;
        k = (k*m)%MOD;
    }
    
    int getIndex(int idx) {
        if (idx >= nums.size())
            return -1;
        ll k_ = (k*quickmul(ks[idx],MOD-2))%MOD;
        ll b_ = (b - (k_*bs[idx])%MOD + MOD)%MOD;
        return (nums[idx]*k_+b_)%MOD;
    }

    int quickmul(int a, int p){
        int res = 1;
        while (p){
            if (p&1)
                res = ((ll)res*a)%MOD;
            a = ((ll)a*a)%MOD;
            p >>= 1;
        }
        return res;
    }
};

/**
 * Your Fancy object will be instantiated and called as such:
 * Fancy* obj = new Fancy();
 * obj->append(val);
 * obj->addAll(inc);
 * obj->multAll(m);
 * int param_4 = obj->getIndex(idx);
 */